Parse the MISC_INFO_5 format in minidumps

Gankra opened the pull request:
Fixes #16

This is a bit messy because I misunderstood the scope of #16 and started working towards actually walking through the XSAVE entries that MISC_INFO_5 describes before realizing that wasn't actually desired at the moment. So this has some extra "useless" definitions that cover what the meaning of the enabled_features bitfield is, but I figured I'd leave it there in case it's ever desirable to properly process that data.

Gankra commented:
Oh also my naming scheme is all over the place because I don't quite get where SCREAMING_CASE and RustCase should be used in this file. Presumably something relating to the official headers and breakpad but MISC_INFO_3+ aren't in the headers so...?

Gankra commented:
Oh also I still need to test this -- need to dig up a minidump that has this data.

jan-auer commented on minidump-common/src/format.rs:
```diff
@@ -1389,33 +1389,74 @@ multi_structs! {
+
+#[derive(Clone, Copy, Pread, SizeWith)]
+pub struct XSTATE_FEATURE {
+    pub offset: u32,
```
It would be great to assume some of Crashpad's doc comments:

- https://crashpad.chromium.org/doxygen/structXSTATE__FEATURE.html
- https://crashpad.chromium.org/doxygen/structXSTATE__CONFIG__FEATURE__MSC__INFO.html

Gankra replied:
I made my own versions since these refer a lot to their own APIs

jan-auer commented on minidump-common/src/format.rs:
```diff
@@ -1389,33 +1389,74 @@ multi_structs! {
+    // TODO?: read xstate_data.enabled_features and process the extra XSAVE
+    // sections that should follow MISC_INFO_5 entries as documented in
+    // Volume 1, Chapter 13 of the "Intel 64 and IA-32 Architectures
+    // Software Developer’s Manual".
```
This is a great idea. 

If I understand @luser's intentions correctly, he prefers to keep these types close to the binary format. Considering this, you could provide helper methods that look into `enabled_features` and then return an `Option<XStateFeature>` by `XStateFeatureIndex`.

gabrielesvelto replied:
We'd like to implement that in a follow-up patch to keep complexity limited and also because we still have to gather a bunch of minidumps to test the feature (I've gathered minidumps  with 8 different combinations of XSTATE flags but they don't cover all of them).

Gankra replied:
let's leave this for future work, since it potentially involves some restructuring of how we read this type out that I need to think through more.

jan-auer replied:
Could you outline your plans? It looks like all you need is one additional method that does roughly this:

```rust
impl XSTATE_CONFIG_FEATURE_MSC_INFO {
    pub fn feature(&self, index: XStateFeatureIndex) -> Option<XSTATE_FEATURE> {
        let index = index as usize;

        XStateFeatures::from_bits(1 << index)
            .filter(|f| self.enabled_features.contains(*f))
            .and_then(|_| self.features.get(index))
            .copied()
    }
}
```

jan-auer reviewed (commented):
Thanks for this patch! 

I agree with your naming concerns, and it seems like you've found good middle grounds that we can go with.

If you rebase or merge `master`, all unrelated clippy and rustc errors should be resolved.

Gankra commented:
I've also added an extra underscore to the MISC_INFO_X names to both match the windows headers and luser's original implementation of this.

jan-auer commented on minidump-common/src/format.rs:
```diff
@@ -1386,37 +1386,91 @@ multi_structs! {
+
+/// Several known entries in `XSTATE_CONFIG_FEATURE_MSC_INFO.features`.
+#[repr(usize)]
+pub enum XStateFeatureIndex {
```
Let's also add the usual trait implementations here.

jan-auer replied:
This still applies, could you add at least `Clone, Copy`, and depending on what you think makes sense also `Ord, Eq, Hash`?

jan-auer reviewed (commented):
Thanks for the updates! This looks great and is G2G once tests have been added. See my one comment below.

luser commented:
I had a WIP patch for this from a while back, I can't remember why I never finished it though: https://github.com/luser/rust-minidump/commit/6708884864c0526aa1909aa92f729738bafc4509

The naming conventions are mostly "match the official Microsoft struct names when possible", but there are a bunch of structs that don't originate from Microsoft headers, so it's more convention than anything.

luser commented on minidump-common/src/format.rs:
```diff
@@ -1386,37 +1386,91 @@ multi_structs! {
+}
+
+// For whatever reason Pread array derives use 0u8.into() instead of Default to
+// create an initial array to write into. Weird.
```
This is probably just "Default not implemented for arrays of arbitrary length".

Gankra commented:
I accidentally got my branches mixed up so the printing patches in my other PR are now folded into this one. I've updated the commit message to reflect the current state.

The new printing output now looks like this:

```
MINIDUMP_MISC_INFO
  size_of_info                 = 1364
  flags1                       = 3f7
  process_id                   = 12144
  process_create_time          = 0x605a4f4d 2021-03-23 20:27:57
  process_user_time            = 350
  process_kernel_time          = 457
  processor_max_mhz            = 3701
  processor_current_mhz        = 3701
  processor_mhz_limit          = 3701
  processor_max_idle_state     = 2
  processor_current_idle_state = 2
  process_integrity_level      = 8192
  process_execute_flags        = 3a
  protected_process            = 0
  time_zone_id                 = 2
  time_zone                    =
    bias          = 480
    standard_name = Pacific Standard Time
    standard_date = 0000-11-01 02:00:00:000000000
    standard_bias = 0
    daylight_name = Pacific Daylight Time
    daylight_date = 0000-03-02 02:00:00:000000000
    daylight_bias = -60
  build_string                 = 19041.1.amd64fre.vb_release.191206-1406
  dbg_bld_str                  = dbgcore.i386,10.0.19041.546
  xstate_data                  =
    feature  0 - LEGACY_FLOATING_POINT :  offset    0, size  160
    feature  1 - LEGACY_SSE            :  offset  160, size  256
    feature  2 - GSSE_AND_AVX          :  offset  576, size  256
  process_cookie               = 11984912
```

Gankra commented:
Oh right: I removed the bitflags impl from the enabled_features field because it didn't really do anything and bitflags isn't designed for situations where unspecified bits can be set. So now it's a bare integer that's interpreted by the iterator impl I added.

Gankra commented:
Apparently you can mess up a rebase so hard github just closes the issue because it thinks your PR was merged lol

Gankra commented:
Is there anything else that needs to be done to push this forward?

jan-auer commented on minidump-common/src/format.rs:
```diff
@@ -1446,37 +1460,133 @@ multi_structs! {
+}
+
+impl<'a> Iterator for XstateFeatureIter<'a> {
+    type Item = (usize, &'a XSTATE_FEATURE);
```
Two questions here:

1. Since `XSTATE_FEATURE` is `Copy`, can you just return the owned instance?
2. The index would always be one of `XStateFeatureIndex`, would it make sense to use that type instead?

Gankra replied:
It can't be an XStateFeatureIndex because there can be unknown indices. We could fill up the enum with Unknown1, Unknown2... but I don't think that's particularly helpful. We *could* yield `Option<XStateFeatureIndex>` but that's a bit weird, and may be limiting if the user of our library has extra info about index mappings.

jan-auer commented on minidump-common/src/format.rs:
```diff
@@ -1446,37 +1460,133 @@ multi_structs! {
+
+/// An iterator of all the enabled features in an XSTATE_CONFIG_FEATURE_MSC_INFO.
+#[derive(Debug)]
+pub struct XstateFeatureIter<'a> {
```
This structure is spelled `Xstate` vs the index enum that is spelled `XState`. Even though it's just a minor difference, it would be great consolidating this.

jan-auer reviewed (approved):
Looks good to me, especially the updates to minidump_dump and SynthMinidump. Noticed three minor points in final review, other than that this is ready to merge.

Thanks!

Gankra commented:
Should be good to go.

jan-auer commented:
Thank you!
